pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/InvertedIndexAndDictionaryBasedForwardIndexCreator.java

```diff
@@ -0,0 +1,609 @@
+      }
+    }
+
+    // Delete the marker file.
+    FileUtils.deleteQuietly(inProgress);
+
+    LOGGER.info("Created a new forward index for segment: {}, column: {}, isTemporary: {}", segmentName, _columnName,
+        _isTemporaryForwardIndex);
+  }
+
+  private Map<String, String> createForwardIndexForSVColumn()
+      throws IOException {
+    try (BitmapInvertedIndexReader invertedIndexReader =
+        (BitmapInvertedIndexReader) LoaderUtils.getInvertedIndexReader(_segmentWriter, _columnMetadata);
+        Dictionary dictionary = LoaderUtils.getDictionary(_segmentWriter, _columnMetadata)) {
+      boolean shouldTrackVarLengthMetadata = !_columnMetadata.getFieldSpec().getDataType().isFixedWidth();
+      int[] lengthOfLongestEntry = shouldTrackVarLengthMetadata ? new int[]{0} : new int[]{-1};
+      // Construct the forward index in the values buffer
+      for (int dictId = 0; dictId < _cardinality; dictId++) {
+        ImmutableRoaringBitmap docIdsBitmap = invertedIndexReader.getDocIds(dictId);
+        int finalDictId = dictId;
+        docIdsBitmap.stream().forEach(docId -> putInt(_forwardIndexValueBuffer, docId, finalDictId));
+        if (shouldTrackVarLengthMetadata) {
+          trackLengthOfLongestEntry(dictionary, lengthOfLongestEntry, dictId);
+        }
+      }
+
+      IndexCreationContext.Forward context =
+          IndexCreationContext.builder().withIndexDir(_segmentMetadata.getIndexDir())
+              .withColumnMetadata(_columnMetadata).withforwardIndexDisabled(false).withDictionary(_dictionaryEnabled)
+              .withLengthOfLongestEntry(lengthOfLongestEntry[0]).build()
+              .forForwardIndex(_chunkCompressionType, _indexLoadingConfig.getColumnProperties());
+
+      writeToForwardIndex(dictionary, context);
+
+      // Setup and return the metadata properties to update
+      Map<String, String> metadataProperties = new HashMap<>();
+      metadataProperties.put(getKeyFor(_columnName, HAS_DICTIONARY), String.valueOf(_dictionaryEnabled));
+      metadataProperties.put(getKeyFor(_columnName, DICTIONARY_ELEMENT_SIZE),
+          String.valueOf(_dictionaryEnabled ? _columnMetadata.getColumnMaxLength() : 0));
+      return metadataProperties;
+    }
+  }
+
+  private Map<String, String> createForwardIndexForMVColumn()
+      throws IOException {
+    try (BitmapInvertedIndexReader invertedIndexReader =
+        (BitmapInvertedIndexReader) LoaderUtils.getInvertedIndexReader(_segmentWriter, _columnMetadata);
+        Dictionary dictionary = LoaderUtils.getDictionary(_segmentWriter, _columnMetadata)) {
+      // Construct the forward index length buffer and create the inverted index values and length buffers
+      int[] maxNumberOfMultiValues = new int[]{0};
+      final boolean shouldTrackVarLengthMetadata = !_columnMetadata.getFieldSpec().getDataType().isFixedWidth();
+      int[] lengthOfLongestEntry = shouldTrackVarLengthMetadata ? new int[]{0} : new int[]{-1};
```

Comment:
instead of tracking the flag in shouldTrackVarLengthMetadata additional variable, we can just simply check if it is fixedWidth or not right ? Something like ....

```
boolean fixedWidth = _columnMetadata.getFieldSpec().getDataType().isFixedWidth()
if (!fixedWidth) {
   trackLengthOfLongestEntry(dictionary, lengthOfLongestEntry, dictId);
}
```

Just trying to see if code can be made less verbose somehow....

Reply:
done